cluster-autoscaler/simulator/drain.go

```diff
@@ -31,66 +30,45 @@ import (
 	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
 )
 
-// GetPodsToMove returns a list of pods that should be moved elsewhere
-// and a list of DaemonSet pods that should be evicted if the node
-// is drained. Raises error if there is an unreplicated pod.
-// Based on kubectl drain code. If listers is nil it makes an assumption that RC, DS, Jobs and RS were deleted
-// along with their pods (no abandoned pods with dangling created-by annotation).
-// If listers is not nil it checks whether RC, DS, Jobs and RS that created these pods
-// still exist.
-// TODO(x13n): Rewrite GetPodsForDeletionOnNodeDrain into a set of DrainabilityRules.
+// GetPodsToMove returns a list of pods that should be moved elsewhere and a
+// list of DaemonSet pods that should be evicted if the node is drained.
+// Raises error if there is an unreplicated pod.
+// Based on kubectl drain code. If listers is nil it makes an assumption that
+// RC, DS, Jobs and RS were deleted along with their pods (no abandoned pods
+// with dangling created-by annotation).
+// If listers is not nil it checks whether RC, DS, Jobs and RS that created
+// these pods still exist.
 func GetPodsToMove(nodeInfo *schedulerframework.NodeInfo, deleteOptions options.NodeDeleteOptions, drainabilityRules rules.Rules, listers kube_util.ListerRegistry, remainingPdbTracker pdb.RemainingPdbTracker, timestamp time.Time) (pods []*apiv1.Pod, daemonSetPods []*apiv1.Pod, blockingPod *drain.BlockingPod, err error) {
-	var drainPods, drainDs []*apiv1.Pod
 	if drainabilityRules == nil {
-		drainabilityRules = rules.Default()
+		drainabilityRules = rules.Default(deleteOptions)
 	}
 	if remainingPdbTracker == nil {
 		remainingPdbTracker = pdb.NewBasicRemainingPdbTracker()
 	}
 	drainCtx := &drainability.DrainContext{
 		RemainingPdbTracker: remainingPdbTracker,
-		DeleteOptions:       deleteOptions,
+		Listers:             listers,
+		Timestamp:           timestamp,
 	}
 	for _, podInfo := range nodeInfo.Pods {
 		pod := podInfo.Pod
 		status := drainabilityRules.Drainable(drainCtx, pod)
 		switch status.Outcome {
-		case drainability.UndefinedOutcome:
-			pods = append(pods, podInfo.Pod)
-		case drainability.DrainOk:
+		case drainability.UndefinedOutcome, drainability.DrainOk:
+			if drain.IsPodLongTerminating(pod, timestamp) {
```

Comment:
No, SkipDrain doesn't get added to the drainable pods list (notice how there is no switch case for SkipDrain).

UndefinedOutcome and DrainOk are treated identically after the refactor.

Reply:
Apologies if I wasn't clear. I didn't mean to treat `DrainOk`/`UndefinedOutcome` and `SkipDrain` the same here, just that the special case for long terminating pods effectively causes `DrainOk` to be treated as if it was `SkipDrain`. Since now `longterminating` rule simply returns `SkipDrain`, the whole special casing is really a no-op.

Reply:
Got it. You want this condition removed. Agreed that this can be delegated to the longterminating rule.